app/services/case_extend_sar_deadline_service.rb

```diff
@@ -0,0 +1,89 @@
+class CaseExtendSARDeadlineService
+  attr_reader :result, :error
+
+  def initialize(user, kase, extension_days, reason)
+    @user = user
+    @case = kase
+    @extension_days = extension_days
+    @extension_deadline = new_extension_deadline(@extension_days.to_i)
+    @reason = reason
+    @result = :incomplete
+  end
+
+  def call
+    ActiveRecord::Base.transaction do
+      if validate_params
+        @case.state_machine.extend_sar_deadline!(
+          acting_user: @user,
+          acting_team: @user.team_for_case(@case),
+          final_deadline: @extension_deadline,
+          original_final_deadline: @case.external_deadline,
+          message: message
+        )
+
+        @case.extend_deadline!(@extension_deadline)
+        @result = :ok
+      end
+    end
+    @result
+  rescue => err
+    Rails.logger.error err.to_s
+    Rails.logger.error err.backtrace.join("\n\t")
+    @error = err
+    @result = :error
+  end
+
+  private
+
+  def new_extension_deadline(extend_by)
+    if @case.deadline_extended?
+      @case.external_deadline + extend_by.days
+    else
+      @case.initial_deadline + extend_by.days
+    end
+  end
+
+  def message
+    "#{@reason}\nDeadline extended by #{@extension_days} days"
+  end
+
+  def validate_params
+    unless @reason.present?
+      @case.errors.add(:reason_for_extending, "can't be blank")
+      @result = :validation_error
+    end
+
+    unless extension_date_valid?
+      @result = :validation_error
+    end
+
+    @result = :ok if @case.errors.empty?
+  end
+
+  def extension_date_valid?
```

Comment:
Both the case and this service know about Settings.sar_extension_limit and what it means (the error message here expresses the same intent as coded in Case#max_allowed_deadline_date) Maybe this is a justification for adding a validation to the case model?

Reply:
Discussed with @starswan - leave as is for now as the handling of validations is a system wide issue requiring a broader discussion